### PR TITLE
Rn location restrictions on mpr report

### PR DESCRIPTION
**Pull request by Rohit25negi**
This function was returning all the locations for the MPR report no matter what permission the user had.
This change limits the locations being returned depending on the user's permissions.

Asana Ticket: https://app.asana.com/0/1187146330310233/1186445439810547

### CODE REVIEW

File: custom/icds_reports/filters.py [hunk 1]
```diff
@@ -66,7 +74,9 @@ def loc_to_json(loc, project):
                 # there are some instances in viewing archived locations where we don't actually
                 # support drilling all the way down.
                 break
-            this_loc['children'] = [loc_to_json(loc, project) for loc in children]
+            this_loc['children'] = [loc_to_json(loc, project) for loc in children if
+                                    user_can_access_location_id(domain, user, loc.location_id) or
```
Comment (mkangia) [inline custom/icds_reports/filters.py:83]: This is going to be a lot of sql queries on each page load

File: custom/icds_reports/filters.py [hunk 2]
```diff
@@ -35,16 +37,22 @@ def loc_to_json(loc, project):
...
+
     project = Domain.get_by_name(domain)
 
     locations = SQLLocation.root_locations(domain)
+    user_location_list = user_locations()
     if not show_test:
         locations = [
-            loc for loc in locations if loc.metadata.get('is_test_location', 'real') != 'test'
+            loc for loc in locations if loc.metadata.get('is_test_location', 'real') != 'test' and
+                                        loc.location_id in user_location_list
```
Comment (mkangia) [inline custom/icds_reports/filters.py]: this would skip the check for accessible locations if we are viewing test locations. I think that is not what we want. This check should be independent of test locations being viewed or not

**Review by mkangia — CHANGES_REQUESTED** [review]
I think the changes should
1. get all accessible location ids at once and then use that do whatever is needed in the method
2. if a user has access to all locations then just skip checking for location access.

File: custom/icds_reports/filters.py (hunk 1 above)
Comment (mkangia) [reply custom/icds_reports/filters.py:83]: Can you help understand what is happening if there is a selected location id, specially `this_loc = [k for k in parent['children'] if k['uuid'] == loc.location_id][0]`?

**Comment by Rohit25negi** [comment]
> I think the changes should
> 
> 1. get all accessible location ids at once and then use that do whatever is needed in the method
> 2. if a user has access to all locations then just skip checking for location access.

That's definitely smarter than how I saw it. Sure, let me inculcate these changes.

**Comment by czue** [comment]
looks like @mkangia did a good review on this already, but let me know if you need more eyes!

File: custom/icds_reports/filters.py (hunk 1 above)
Comment (Rohit25negi) [reply custom/icds_reports/filters.py]: hey yes,
selected ID comes as id of location assigned to the user.
 `this_loc = [k for k in parent['children'] if k['uuid'] == loc.location_id][0]` and the outer loop is actually preparing a dict in tree like structure: 

```

[{
name:<>,
.
.,
children:[
{
name:<>,
children:[...]
}
],
}]

```
the outer loop is applied on the selected_location ancestor hierarchy  only and skips other parent locations

Comment (Rohit25negi) [reply custom/icds_reports/filters.py]: I think the thing it is trying to achieve could have been done with recursion. But Not sure why this way was chosen.

**Comment by Rohit25negi** [comment]
@mkangia  its ready for second round of review

File: custom/icds_reports/filters.py [hunk 3]
```diff
@@ -38,12 +38,22 @@ def loc_to_json(loc, project):
...
 
     locations = SQLLocation.root_locations(domain)
+
+    if user.has_permission(domain, 'access_all_locations'):
+        all_user_locations = []
+    else:
+        user_location_list = SQLLocation.objects.accessible_to_user(domain, user)
+        user_location_list = [loc.location_id for loc in user_location_list]
+        user_location_ancestors = [loc.location_id for loc in user.sql_location.get_ancestors(include_self=True)]
+        all_user_locations = user_location_list + user_location_ancestors
```
Comment (mkangia) [inline custom/icds_reports/filters.py:48]: I think this can be further improved. Suggesting something that should be tested/confirmed
```
user_has_access_to_all_locations = user.has_permission(domain, 'access_all_locations')
if user_has_access_to_all_locations:
    accessible_root_locations = SQLLocation.root_locations(domain)
else:
    accessible_root_locations = user.sql_location.get_ancestors_of_type(type_code='state')
    
    # ...

    loc_json = [
        loc_to_json(loc, project) for loc in accessible_root_locations
    ]

# ....
if selected_location_id:
    if not user_has_access_to_all_locations:
        primary_location = user.sql_location
        accessible_ancestor_ids = primary_location.get_ancestors().values_list('location_id', flat=True)
        accessible_descendant_ids = primary_location.get_descendants().values_list('location_id', flat=True)
        accessible_location_ids = set(primary_location.location_id + accessible_ancestor_ids + accessible_descendant_ids)
```

File: custom/icds_reports/filters.py (hunk 1 above)
Comment (mkangia) [reply custom/icds_reports/filters.py]: >  selected ID comes as id of location assigned to the user.

can you confirm that it's not the location selected in the view?

File: custom/icds_reports/filters.py (hunk 3 above)
Comment (Rohit25negi) [reply custom/icds_reports/filters.py]: incorporated

File: custom/icds_reports/filters.py [hunk 4]
```diff
@@ -66,9 +78,11 @@ def loc_to_json(loc, project):
                 # there are some instances in viewing archived locations where we don't actually
                 # support drilling all the way down.
                 break
-            this_loc['children'] = [loc_to_json(loc, project) for loc in children]
```
Comment (mkangia) [inline custom/icds_reports/filters.py:69]: just spent roughly 30 mins trying to understand what is happening here, so going to put this down as suggestion to consider at least
my understanding is that this block just wants to find the children locations of all ancestors
```
selected_location = SQLLocation.active_objects.get(domain=domain, location_id=selected_loc_id)
json_at_level = loc_json  # json at level at which we should find the ancestor
for ancestor in selected_location.get_ancestors():  # this would start with top level ancestor first
    ancestor_loc_dict_in_json = None
    for loc_dict in json_at_level:
        if loc_dict['uuid'] == ancestor.location_id:
            ancestor_loc_dict_in_json = loc_dict
            break
    if not ancestor_loc_dict_in_json:  # could not find the ancestor at the level, user should not have reached at tihs point to try access an ancestor not permitted access to
        break
    if self.user.has_permission(domain, 'access_all_locations'):
        ancestor_loc_dict_in_json['children'] = [loc_to_json(loc, project) for loc in ancestor.child_locations()
    else:
        ancestor_loc_dict_in_json['children'] = [loc_to_json(loc, project) for loc in ancestor.child_locations()
                                                                      if loc.location_id in accessible_location_ids]
    json_at_level = ancestor_loc_dict_in_json['children']  # reset level to one level down to find next ancestor
```

File: custom/icds_reports/filters.py [hunk 5]
```diff
@@ -37,38 +37,63 @@ def loc_to_json(loc, project):
...
+
+            # could not find the ancestor at the level,
+            # user should not have reached at this point to try access an ancestor not permitted access to
+            if ancestor_loc_dict_in_json is None:
                 break
-            this_loc['children'] = [loc_to_json(loc, project) for loc in children]
-            parent = this_loc
 
+            children = ancestor.child_locations()
+            if user.has_permission(domain, 'access_all_locations'):
```
Comment (mkangia) [inline custom/icds_reports/filters.py:85]: since this is called multiple times, its definitely getting this once on top

File: custom/icds_reports/filters.py [hunk 6]
```diff
@@ -37,38 +37,63 @@ def loc_to_json(loc, project):
...
+                                     if loc.metadata.get('is_test_location', 'real') != 'test']
 
-    loc_json = [loc_to_json(loc, project) for loc in locations]
+    loc_json = [loc_to_json(loc, project) for loc in accessible_root_locations]
 
     # if a location is selected, we need to pre-populate its location hierarchy
     # so that the data is available client-side to pre-populate the drop-downs
     if selected_loc_id:
+        if not user.has_permission(domain, 'access_all_locations'):
+            primary_location = user.sql_location
```
Comment (mkangia) [inline custom/icds_reports/filters.py:58]: `user.sql_location` is not a cached property, not sure why though. So this would fire up a query each time, Fetching this on top once would be nice.

File: custom/icds_reports/filters.py [hunk 7]
```diff
@@ -37,38 +37,63 @@ def loc_to_json(loc, project):
...
     project = Domain.get_by_name(domain)
 
-    locations = SQLLocation.root_locations(domain)
+    if user.has_permission(domain, 'access_all_locations'):
+        accessible_root_locations = SQLLocation.root_locations(domain)
+    else:
+        accessible_root_locations = [(user.sql_location
+                                      .get_ancestors(include_self=True)
+                                      .get(location_type__code='state'))
+                                     ]
```
Comment (mkangia) [inline custom/icds_reports/filters.py:46]: nit: if the sql_location of the user is a root location on its own, this basically does not getting anything new i.e the case where the sql_location is itself the top most location, is implicitly handled here. One can skip getting the ancestors if the sql_location is itself the root, but might not be worth. (just noted this down as a observation, there are pros and cons to both implementations)

```
user_primary_location = user.sql_location
if user_primary_location.parent_id is None:
    accessible_root_locations = [user_primary_location]
else:
    accessible_root_locations = [(user.sql_location.get_ancestors().filter(parent_id__isnull=True)]
```

But worth considering `parent_id__isnull=True`, in case `state` is not the topmost location anymore, sometime in future though that will break many other things.

File: custom/icds_reports/filters.py [hunk 8]
```diff
@@ -37,38 +37,63 @@ def loc_to_json(loc, project):
...
+
     if not show_test:
-        locations = [
-            loc for loc in locations if loc.metadata.get('is_test_location', 'real') != 'test'
-        ]
+        accessible_root_locations = [loc for loc in accessible_root_locations
+                                     if loc.metadata.get('is_test_location', 'real') != 'test']
```
Comment (mkangia) [inline custom/icds_reports/filters.py:50]: looks like a bug in the original implementation that filters out for test root locations, but does not do that again if any of the child location is marked as test for a non-test state. Not saying you should fix it here but does look like something that can go wrong.

File: custom/icds_reports/filters.py [hunk 9]
```diff
@@ -37,38 +37,63 @@ def loc_to_json(loc, project):
...
     if selected_loc_id:
+        if not user.has_permission(domain, 'access_all_locations'):
+            primary_location = user.sql_location
+            accessible_ancestor_ids = primary_location.get_ancestors().values_list('location_id', flat=True)
+            accessible_descendant_ids = primary_location.get_descendants().values_list('location_id', flat=True)
+            accessible_location_ids = set(
+                [primary_location.location_id] + list(accessible_ancestor_ids) + list(accessible_descendant_ids))
+        else:
+            accessible_location_ids = []
+
```
Comment (mkangia) [inline custom/icds_reports/filters.py:65]: if it looks better,
```
skip_access_check = user.has_permission(domain, 'access_all_locations')
if not skip_access_check:
    accessible_location_ids = _get_accessible_location_ids(user_primary_location)
```
```
def _get_accessible_location_ids(user_primary_location):
    accessible_location_ids = set([primary_location.location_id])
    accessible_location_ids.update(primary_location.get_ancestors().values_list('location_id', flat=True))
    accessible_location_ids.update(primary_location.get_descendants().values_list('location_id', flat=True))
    return accessible_location_ids
```

File: custom/icds_reports/filters.py [hunk 10]
```diff
@@ -37,38 +37,63 @@ def loc_to_json(loc, project):
...
-                # if we couldn't find this location the view just break out of the loop.
-                # there are some instances in viewing archived locations where we don't actually
-                # support drilling all the way down.
+        json_at_level = loc_json  # json at level at which we should find the ancestor
+        for ancestor in selected.get_ancestors():  # this would start with top level ancestor first
+            ancestor_loc_dict_in_json = None
+            for parent_location in json_at_level:
+                if parent_location['uuid'] == ancestor.location_id:
+                    ancestor_loc_dict_in_json = parent_location
+                    break
```
Comment (mkangia) [inline custom/icds_reports/filters.py:77]: nit: can actually extract this out as a method which would make it easier to understand this whole big method

```
for ancestor in selected.get_ancestors()
    ancestor_loc_dict_in_json = _get_ancestor_loc_dict(json_at_level)
```

```
def _get_ancestor_loc_dict(in_json):
    for loc_dict in in_json:
        if loc_dict['uuid'] == ancestor.location_id:
            return loc_dict
    return None
```

File: custom/icds_reports/filters.py [hunk 11]
```diff
@@ -37,38 +37,63 @@ def loc_to_json(loc, project):
...
+
+            # could not find the ancestor at the level,
+            # user should not have reached at this point to try access an ancestor not permitted access to
```
Comment (mkangia) [inline custom/icds_reports/filters.py:80]: nit: user should not have reached at this point to try and access an ancestor that is not permitted

File: custom/icds_reports/filters.py [hunk 12]
```diff
@@ -37,38 +37,63 @@ def loc_to_json(loc, project):
...
-        for loc in lineage:
-            children = loc.child_locations()
-            # find existing entry in the json tree that corresponds to this loc
-            try:
-                this_loc = [k for k in parent['children'] if k['uuid'] == loc.location_id][0]
-            except IndexError:
-                # if we couldn't find this location the view just break out of the loop.
-                # there are some instances in viewing archived locations where we don't actually
-                # support drilling all the way down.
+        json_at_level = loc_json  # json at level at which we should find the ancestor
```
Comment (mkangia) [inline custom/icds_reports/filters.py:71]: json in which we should find the ancestor in iteration

File: custom/icds_reports/filters.py [hunk 13]
```diff
@@ -37,38 +37,63 @@ def loc_to_json(loc, project):
...
+                ancestor_loc_dict_in_json['children'] = [
+                    loc_to_json(loc, project) for loc in children
+                ]
+            else:
+                ancestor_loc_dict_in_json['children'] = [
+                    loc_to_json(loc, project) for loc in children
+                    if loc.location_id in accessible_location_ids
+                ]
+
+            # reset level to one level down to find next ancestor
```
Comment (mkangia) [inline custom/icds_reports/filters.py:95]: reset level to one level down to find ancestor in next iteration

**Review by mkangia — APPROVED** [review]
Mostly looking good. The suggestions are improvements that should surely be considered (and tested)

Thanks for your patience with this @Rohit25negi . This method took me at least an hour to truly understand, if not more, the effort was just to try and make sure no one else needs to.

**Comment by mkangia** [comment]
Once the suggestions are concluded, should be good to @czue to give it a look

File: custom/icds_reports/filters.py (hunk 8 above)
Comment (Rohit25negi) [reply custom/icds_reports/filters.py:50]: hey, I think this is doing similarly we are doing in the dashboard and might have spun off from the thinking of the system that no lower-level location is marked as a test if its state is nontest.
Although this is definitely just a functional level agreement which is definitely not restricted by any UI. 

I think I will just use this opportunity to fix this as well and make this check on children as well.

File: custom/icds_reports/filters.py (hunk 6 above)
Comment (Rohit25negi) [reply custom/icds_reports/filters.py:58]: make sense

**Comment by Rohit25negi** [comment]
suggestions are implemented. @czue  does it look good to you?

PS: I have tested it on Stage, Looking to deploy on TCL today.

**Review by mkangia — APPROVED** [review]
Thanks for taking care of this @Rohit25negi 
(just saw the duplicate code between this method and load_locs_json in HQ)

File: custom/icds_reports/filters.py [hunk 14]
```diff
@@ -47,7 +47,7 @@ def _get_ancestor_loc_dict(location_list, location):
                 return parent_location
         return None
 
-    user_location = user.sql_location
+    user_location = user.get_sql_location(domain)
```
Comment (mkangia) [inline custom/icds_reports/filters.py:50]: i thought all dashboard users were mobile users, thanks for catching that.

File: custom/icds_reports/filters.py [hunk 15]
```diff
@@ -85,7 +85,7 @@ def _get_ancestor_loc_dict(location_list, location):
             if ancestor_loc_dict_in_json is None:
                 break
 
-            child_locations = ancestor.child_locations(include_archive_ancestors=True)
+            child_locations = ancestor.child_locations()
```
Comment (mkangia) [inline custom/icds_reports/filters.py:88]: why this change?

Comment (Rohit25negi) [reply custom/icds_reports/filters.py:88]: This was making it only return archieve locations.

Comment (mkangia) [reply custom/icds_reports/filters.py:88]: i am confused. That says "include archived ancestors", you mean you didn't want archived ancestors and just child locations? if yes, that sounds good

Comment (Rohit25negi) [reply custom/icds_reports/filters.py]: offlined and changed it to get_children()

**Review by calellowitz — CHANGES_REQUESTED** [review]
It is highly likely that I am missing something, but this feels like a very complicated way of basically just getting the locations a user has access to, above the selected one, which seems like the kind of thing that can be done in one query and with minimal transform logic on the results. That does get a bit more complicated with users who have access to all locations, but even that shouldnt be too bad. This looks very similar to [the location ancestor view](https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/views.py#L727) in the dashboard. That view is definitely not the clearest but does seem both more concise and efficient than this one

File: custom/icds_reports/filters.py [hunk 16]
```diff
@@ -35,40 +35,66 @@ def loc_to_json(loc, project):
...
+        accessible_location_ids.update(primary_location.get_descendants().values_list('location_id', flat=True))
+        return accessible_location_ids
+
+    def _get_ancestor_loc_dict(location_list, location):
+        for parent_location in location_list:
+            if parent_location['uuid'] == location.location_id:
+                return parent_location
+        return None
+
+    user_location = user.get_sql_location(domain)
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds_reports/filters.py:50]: hmm, I thought we wanted to just check for the primary location of the user. Will let @Rohit25negi share why he used that. That was the one reason we didn't end up using `accessible_to_user` method because that uses assigned location and not primary location and ended up finding out the accessible locations ourselves.

File: custom/icds_reports/filters.py [hunk 17]
```diff
@@ -35,40 +35,66 @@ def loc_to_json(loc, project):
...
+    user_has_all_location_access = user.has_permission(domain, 'access_all_locations')
+
+    if user_has_all_location_access:
+        accessible_root_locations = SQLLocation.root_locations(domain)
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds_reports/filters.py:54]: Since we just need the root locations if there is no selected location id, this is just trying to start with that. `accessible_location_ids` is called later only if it is needed. Does that make sense?

File: custom/icds_reports/filters.py [hunk 18]
```diff
@@ -35,40 +35,66 @@ def loc_to_json(loc, project):
...
                 break
-            this_loc['children'] = [loc_to_json(loc, project) for loc in children]
-            parent = this_loc
 
+            child_locations = ancestor.get_children()
+
+            ancestor_loc_dict_in_json['children'] = [
+                loc_to_json(loc) for loc in child_locations
+                if loc.metadata.get('is_test_location', 'real') != 'test'
+                and (user_has_all_location_access or loc.location_id in accessible_location_ids)
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds_reports/filters.py:93]: I am not sure if that would be that straight forward, but probably would be clear on giving it a shot.

but just weighing in, If we change that method to get accessible locations instead of just ids, we would first need to set some structure to that, group them by levels or something and then again do some manipulation to filter out what is relevant for the current search to return. If we succeed at that, can't say if it will still be better.
Additionally, It's possible we won't even need all the locations fetched, for ex: if a user has access to a state, we would be fetching all locations under that state, and currently if the user is selecting say a district, we don't really need location data below that district. 
Additionally, Location search has been reported to be a pain on ICDS and there is a ticket to work on that, 
it slows down extremely, at

https://github.com/dimagi/commcare-hq/blob/8f913988b5c0d1c2e8092c2ea2d1d6915a18c0d5/corehq/apps/locations/models.py#L268-L272

so fetching data for all assigned locations is heavy on ICDS, so would be worth avoiding it. i guess what we are doing here is faster.

this method is quite a duplicate of `load_locs_json` in `locations/utils` which follows the same pattern, so if this an improvement, would be nice to extend it to that as well.

File: custom/icds_reports/filters.py [hunk 19]
```diff
@@ -35,40 +35,66 @@ def loc_to_json(loc, project):
...
-    locations = SQLLocation.root_locations(domain)
     if not show_test:
-        locations = [
-            loc for loc in locations if loc.metadata.get('is_test_location', 'real') != 'test'
-        ]
+        accessible_root_locations = [loc for loc in accessible_root_locations
+                                     if loc.metadata.get('is_test_location', 'real') != 'test']
 
-    loc_json = [loc_to_json(loc, project) for loc in locations]
+    loc_json = [loc_to_json(loc) for loc in accessible_root_locations]
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds_reports/filters.py:64]: the method is called `load_locs_json`. i think that is where this confusion begins.

File: custom/icds_reports/filters.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds_reports/filters.py:50]: Now i remember, I think i got to believe that because of https://github.com/dimagi/commcare-hq/pull/28290/files#r467858278

**Comment by Rohit25negi** [comment]
offlined with Cal. 
Hotfixing this as this is not making the current code worse and people are expecting this to go live today.
I will be working on the change request today.

**Comment by Rohit25negi** [comment]
do not merge, because the icds code is split into a separate repo. But We can still continue review here because it looks still persists the correct diff

File: custom/icds_reports/filters.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/filters.py:54]: Agree with Manish.

File: custom/icds_reports/filters.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/filters.py:50]: oh that's a good point. Do you think its worth doing that? because On a real basis it never happens I think and national level user anyway has national-level access which we are already handling it here.?

File: custom/icds_reports/filters.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/filters.py:93]: agree with Manish

File: custom/icds_reports/filters.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/filters.py:64]: Changing naming

File: custom/icds_reports/filters.py [hunk 20]
```diff
@@ -35,40 +35,66 @@ def loc_to_json(loc, project):
...
-            # find existing entry in the json tree that corresponds to this loc
-            try:
-                this_loc = [k for k in parent['children'] if k['uuid'] == loc.location_id][0]
-            except IndexError:
-                # if we couldn't find this location the view just break out of the loop.
-                # there are some instances in viewing archived locations where we don't actually
-                # support drilling all the way down.
+        json_at_level = loc_json  # json in which we should find the ancestor in iteration
+        for ancestor in selected.get_ancestors():  # this would start with top level ancestor first
+            ancestor_loc_dict_in_json = _get_ancestor_loc_dict(json_at_level, ancestor)
```
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/filters.py:81]: fixing.

File: custom/icds_reports/filters.py [hunk 21]
```diff
@@ -35,40 +35,66 @@ def loc_to_json(loc, project):
...
+            if ancestor_loc_dict_in_json is None:
                 break
-            this_loc['children'] = [loc_to_json(loc, project) for loc in children]
-            parent = this_loc
 
+            child_locations = ancestor.get_children()
+
+            ancestor_loc_dict_in_json['children'] = [
+                loc_to_json(loc) for loc in child_locations
+                if loc.metadata.get('is_test_location', 'real') != 'test'
```
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/filters.py:92]: sure fixing it

File: custom/icds_reports/filters.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/filters.py]: Nvm, I think it's worth adding that support in case. Done making that change.

**Comment by Rohit25negi** [comment]
@calellowitz  could you please take a look. if looks good, I can raise the copy PR on the other commcare-icds repo and get it merged so that it can be deployed tonight.

File: custom/icds_reports/filters.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (calellowitz) [reply custom/icds_reports/filters.py:54]: that makes sense

**Comment by Rohit25negi** [comment]
I have analysed the accessible_to_user function it alone does not looks like slow. So, I am now trying to simplify this function

**Comment by Rohit25negi** [comment]
Addressed the comments and create a new Pr in the new repo : https://github.com/dimagi/commcare-icds/pull/32

**Comment by calellowitz** [comment]
This should be closed right?
